### PR TITLE
[MM-25435] Emoji picker + skin tones

### PR DESCRIPTION
**Pull request by Willyfrog**
#### Summary

Redesigned emoji picker + skin tone selector

it is not finished, but functionality is there, I have yet to go over the design files again and see what I've missed. but I'd like a first review to see if there is anything that needs to change

##### Known issues:
- it takes a bit to load the first time, faster afterwards
- some emojis seems to not be properly clasified as "skinnable", I have to go over those and see if they have any differences from the others (e.g: family emojis)

#### Ticket Link

[MM-25435](https://mattermost.atlassian.net/browse/MM-25435)

#### Screenshots

Screen Shot 2021-06-18 at 15 35 56

#### Release Note

```release-note
Redesigned Emoji Picker & skin selector
```

### CODE REVIEW

**Comment by esethna** [comment]
Thanks @Willyfrog! Looking good, a few things I noticed:

1. If you select a few different default skin tones then execute a search it seems to return a bunch of different options for each skin tone:
image

2. Didn't test this but just confirming, the "Custom Emoji" button will only appear if the user has permissions to access the custom emoji page, correct?

@matthewbirtch can you help with a priority review on this?

**Review by esethna — APPROVED** [review]
Approving as none of the issues I listed are blockers

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Thanks @Willyfrog. This is a good start. I did a first pass review and there are definitely a number of tweaks to bring this up to the spec with the new emoji picker design. I'll try to outline it as clearly as I can. There may be a few more things, but this is all I had time for today, so I thought I would get this out to you.

### Emoji Picker Positioning
**From Message Input**
- Align the right edge of the popover with the right edge of the button that triggers it
- The bottom of the popover should be 4px up from the button that triggers it
image

**From the Message Context**
- Align the *left* edge of the emoji picker with the left edge of the button that triggers it
- If the emoji picker displays above the reaction button, it should be 4px above the top of the button that triggers it
- If the emoji picker displays below the reaction button, it should be 4px below the bottom of the button that triggers it
image

**From the Message Hover Actions Context**
- Align the right edge of the emoji picker with the right edge of the button that triggers it
- If the emoji picker displays above the reaction button, it should be 4px above the top of the button that triggers it
- If the emoji picker displays below the reaction button, it should be 4px below the bottom of the button that triggers it

### UI tweaks

**Main Container**
- The popover container should have a `box-shadow: 0px 8px 24px rgba(0, 0, 0, 0.12);` on it (also known as "Elevation 4" in our design system)
- The Container's radius should be `border-radius: 4px;`

**Tabs**
- The `font-size` for 'Emojis' should be 12px and the `font-weight` should be 600
- There needs to be a divider line directly under the Emojis/gif Tabs
- The first tab should be 12px from the left edge. It looks to be 5px right now

**Category Icons**
- The category icon buttons should have `width: 28px;` and `height: 28px;`
- The icons within the buttons should have `font-size: 18px;`
- The active category icon should be `color: var(--button-bg);`. It looks like it's getting the channel color instead for some reason
- For the hover state of the category icon buttons, the icons should be `color: var(--center-channel-color-72);`
}

**Search and Skin Tone Area**
- Can we use the compass `icon-magnify` instead of the font awesome icon for the search input?
- The search input should only have the blue outlined in the focused state
- The skin tone button should have a hover state. The [design file here](https://www.figma.com/file/GPTxUI7ACBbbdRfM8Dbtl6/MM-25435-Emoji-Picker-UI-Refresh?node-id=13%3A2667) shows the hover state. Sample below as well
image
- The skin tone button's tooltip is displaying behind the popover
- The words `Default skin tone` should have `line-height: 14px;`
- When all the skin tones are showing, each of the skin tone buttons should have a hover state as demonstrated in the [design here](https://www.figma.com/file/GPTxUI7ACBbbdRfM8Dbtl6/MM-25435-Emoji-Picker-UI-Refresh?node-id=13%3A3794)
- It seems like it doesn't always choose the skin tone that I've clicked (see gif below).

- Is it possible to animate the skin tones in (slide from right to left) as they are shown and out (slide from right to left) as they are hidden?

**Emojis List**
- The emoji picker list should have `padding: 4px 0 4px 12px;`
- With the above, you'll need to adjust the category headers and remove the -6px left position
- Each emoji in the list should have container that is `width:36px;` and `height: 36px;` and no margin between items
- Design had 9 emojis in each row

**Footer**
- The text label is misaligned vertically in the footer
image
  - Everything should line up like this:
image
- The `Custom Emoji` button should have `font-weight: 600` and `font-size: 12px;`

**Mobile Webview**
- This will need some work, but I didn't have time to go through it in detail yet. Let's address what we can above first.

**Comment by Willyfrog** [comment]
@matthewbirtch thanks for the detailed review.

for the positioning, if it goes within the text input box usually is because the user resized the window after displaying the emoji picker or the window doesn't fit the emoji picker. I haven't done anything to the positioning of the emoji picker, but I agree it should improve. Would it be ok to add a ticket to take a look at it?

I've started working on the rest of the items, the only one that worries me a bit is the missalignment of the text name of the emoji. when there is more than 1 alias it displays in two lines, if I center the label, the second one gets cut. I'll try to take another look in the morning with fresher eyes.

**Comment by matthewbirtch** [comment]
Thanks @Willyfrog. I'm totally ok with a follow up ticket for the positioning. We can leave that out of this PR if you'd prefer

File: actions/emoji_actions.js [hunk 1]
```diff
@@ -43,6 +43,16 @@ export function incrementEmojiPickerPage() {
     };
 }
 
+export function setRecentSkin(skin) {
+    return async (dispatch) => {
```
Comment (jwilander) [inline actions/emoji_actions.js:47]: Does this need to be async?

**Review by jwilander — APPROVED** [review]
LGTM - just one non-blocking comment

**Comment by jwilander** [comment]
@jgilliam17 are you able to prioritize this today?

File: actions/emoji_actions.js (hunk 1 above)
Comment (Willyfrog) [reply actions/emoji_actions.js:47]: no need, removing `async`

**Comment by Willyfrog** [comment]
> It seems like it doesn't always choose the skin tone that I've clicked (see gif below).

I can't repro, but I made some changes so I might have fixed it without noticing

@matthewbirtch  I think I addressed all your points minus the positioning and the animation, I'll create a couple of tickets for those.

> Didn't test this but just confirming, the "Custom Emoji" button will only appear if the user has permissions to access the custom emoji page, correct?

@esethna  It checks if the custom emoji is enabled in the server, I didn't look for specific permissions for users. I'll fix that one now.

>  The autocomplete should use the default skin tone selected in the picker rather than showing all the options:

I'll add a ticket for this.

**Comment by matthewbirtch** [comment]
> @matthewbirtch I think I addressed all your points minus the positioning and the animation, I'll create a couple of tickets for those.

Thanks @Willyfrog. Looks like the majority are resolved. I'm still seeing the following though. If you'd prefer to address these in a follow-up ticket as well, i'd be ok with that since it's so close.

- Can we shift the tabs to the left a little but to align with the search input below it?
Screen Shot 2021-06-21 at 9 54 32 AM

- There appears to be too much padding on the left and right side of the tabs. Can you make sure the left and right padding are 12px?

- The active category icon should be `color: var(--button-bg);`. It looks like it's getting the channel color instead. 
image

- When I click on a category, it seems to be focusing on the search input for some reason?

- Can we add 4px of margin between the magnify icon and the Placeholder copy on the search input?

- The text and the icon within the search input seems to shift when clicking in and out of the input field. It shouldn't jump around

- The Skin Tone button is getting cut off. Also, the hover state for this button should be rounded square, but it's a rectangle shape right now. 
image

- In the footer, the Emoji is aligned too low and not vertically centered with the rest of the content still.
Screen Shot 2021-06-21 at 9 57 43 AM

- The gif picker layout will need to be updated as well. The design proposed a few updates. Can you create a follow-up ticket for that as well?

**Comment by Willyfrog** [comment]
> When I click on a category, it seems to be focusing on the search input for some reason?

removed focus, but it's been working like that for a while now.

>  The gif picker layout will need to be updated as well. The design proposed a few updates. Can you create a follow-up ticket for that as well?

sure!

**Comment by Willyfrog** [comment]
Added tickets:

[emoji search + skin tones](https://mattermost.atlassian.net/browse/MM-36598)
[emoji picker positioning](https://mattermost.atlassian.net/browse/MM-36599)
[skin tone animation](https://mattermost.atlassian.net/browse/MM-36600)

**Review by matthewbirtch — APPROVED** [review]
Thanks for making all those changes @Willyfrog! It's looking great.

**Comment by matthewbirtch** [comment]
@Willyfrog  We should create another ticket for mobile webview.

image

**Comment by esethna** [comment]
@hmhealey @jgilliam17 can you guys help prioritize a review on this PR and help with merging today?

**Comment by esethna** [comment]
@Willyfrog on the latest test server it looks like you have to select the skin tone twice before it takes effect in the picker

**Comment by jgilliam17** [comment]
( forgot to post this before my meetings)
Thanks @Willyfrog 
Beside what Matt pointed out and now Eric, I noticed that skin tone selector needs two clicks to accept the change when opening from "Add a reaction" +, not the main post emoji picker.

**Comment by Willyfrog** [comment]
@jgilliam17 @esethna I found the problem, it was storing wrong the key in the state so it took extra long to retrieve it from the localstorage (which is meant to persist across sessions, not for updating the component)

File: components/emoji_picker/emoji_picker.jsx [hunk 2]
```diff
@@ -481,7 +489,12 @@ export default class EmojiPicker extends React.PureComponent {
         }
     }
 
-    getCategoryByIndex(index) {
+    onSkinSelected = (skin) => {
+        this.props.actions.setRecentSkin(skin);
+        this.props.actions.persistRecentSkin(skin);
```
Comment (hmhealey) [inline components/emoji_picker/emoji_picker.jsx:494]: Also, moving the user's current skin to preferences means we don't need to explicitly persist it

**Review by hmhealey — CHANGES_REQUESTED** [review]

File: actions/emoji_actions.js [hunk 3]
```diff
@@ -43,6 +43,16 @@ export function incrementEmojiPickerPage() {
     };
 }
 
+export function setRecentSkin(skin) {
```
[the remark this answers is not in the document]
Comment (Willyfrog) [reply actions/emoji_actions.js:46]: it is being saved in the local storage, along with the recent emojis, so it persists between sessions but not across devices

**Comment by Willyfrog** [comment]
> @Willyfrog We should create another ticket for mobile webview.
> 
> image

https://mattermost.atlassian.net/browse/MM-36613 feel free to add any details, I've set it to triage so we can probably discuss it next sprint planning

**Review by jgilliam17 — APPROVED** [review]
Thanks @Willyfrog 
Tested, looks good to merge.
- Verified re-designed emoji picker and skin tone selector
- The issue with 2 clicks to change the skin tone has been resolved. 
@hmhealey Can session issue you raised be addressed in a follow up PR?

**Review by hmhealey — APPROVED** [review]
We're good to switch to Preferences in a separate PR

**Pull request merged**